<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Выводы" data-toc-modified-id="Выводы-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

**Цель проекта:** необходимо разработать метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Данный метод будет использован для защиты персональных данных клиентов страховой компании. Преобразование данных по разработанному методу не должно приводить к ухудшению качества моделей машиннного обучения.

**Описание данных:**
* Признаки: пол, возраст и зарплата застрахованного, количество членов его семьи.
* Целевой признак: количество страховых выплат клиенту за последние 5 лет.

## Загрузка данных

Загружаем необходимые библиотеки.

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

Загружаем файл и сохраняем его содержимое под именем data. Затем выводим основную информацию об исходных данных.

In [2]:
data = pd.read_csv("/datasets/insurance.csv")
data.info()
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
Пол                  5000 non-null int64
Возраст              5000 non-null float64
Зарплата             5000 non-null float64
Члены семьи          5000 non-null int64
Страховые выплаты    5000 non-null int64
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [3]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

Пропусков не выявлено. Изменим тип данных для возраста с "вещественного" на "целочисленный".

In [4]:
data["Возраст"] = data["Возраст"].astype("int64")

Подсчитаем количество полных дубликатов.

In [5]:
data.duplicated().sum()

153

Выявили существенное количество полных дубликатов - удалим их. И для контроля получим основную информацию о датасете.

In [6]:
data.drop_duplicates(inplace=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 5 columns):
Пол                  4847 non-null int64
Возраст              4847 non-null int64
Зарплата             4847 non-null float64
Члены семьи          4847 non-null int64
Страховые выплаты    4847 non-null int64
dtypes: float64(1), int64(4)
memory usage: 227.2 KB


### Выводы 

1. Исходные данные загружены и изучены. Размер исходного датасета: 5000x5.
2. Выявили и удалили 153 полных дубликата.
3. Изменили тип данных для столбца "Возраст" с вещественного на целочисленный.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** При умножении матрицы признаков на обратимую матрицу качество линейной регрессии не изменится.

**Обоснование:** Умножим матрицу признаков на обратимую матрицу $A$ и посмотрим как изменится при этом вектор весов:

$w_{new} = ((XA)^T XA)^{-1} (XA)^T y = (A^T X^T XA)^{-1} A^T X^T y = (A^T (X^T XA))^{-1} A^T X^T y = (X^T XA)^{-1} (A^T)^{-1} A^T X^T y = (X^T XA)^{-1} E X^T y = ((X^T X)A)^{-1} X^T y = A^{-1} (X^T X)^{-1} X^T y = A^{-1} w$

При выполнении преобразований для вычисления вектора весов $w_{new}$ использовались следующие свойства матриц:

$$
(AB)C = A(BC)
$$

$$
(AB)^T = B^T A^T
$$

$$
(AB)^{-1} = B^{-1} A^{-1}
$$

$$
A^{-1} A = E
$$

$$
A E = A
$$

Как видно из полученного результата, вектор весов $w_{new}$ может быть получен умножением матрицы $A^{-1}$ на исходный вектор весов $w$

Вычислим вектор предсказаний $a_{new}$ для рассматриваемого случая, когда матрица признаков $X$ умножена на обратимую матрицу $A$:

$a_{new} = XAw_{new} = XAA^{-1}w = XEw = Xw = a$

Таким образом, вектор предсказаний не изменился. Качество модели определяется вектором целевых признаков (y) и вектором предсказаний (a). Так как данные вектора остались без изменений, то качество модели также не поменялось.

### Вывод 

При умножении матрицы признаков на обратимую матрицу качество линейной регрессии не изменится, так как не изменится вектор предсказаний. При этом вектор весов изменится.

## Алгоритм преобразования

Для защиты данных можно предложить следующий алгоритм преобразования данных. Матрицу признаков необходимо умножить на обратимую матрицу. Это не повлечет за собой изменения предсказаний и качества линейной регрессии, что было доказано в разделе 2. 

## Проверка алгоритма

Подготовим исходные данные для машинного обучения

In [7]:
features = data.drop("Страховые выплаты", axis=1)
target = data["Страховые выплаты"]
numeric = ["Возраст", "Зарплата", "Члены семьи"]
scaler = StandardScaler()
scaler.fit(features[numeric])
features[numeric] = scaler.transform(features[numeric])

Создадим обратимую матрицу. Обратимая матрица - квадратная матрица, кроме того из условия возможности умножения матрицы признаков X на обратимую матрицу, количество строк в последней должно быть равно четырем (количество столбцов в матрице признаков). Таким образом размер обратимой матрицы строго фиксирован и равен 4х4.

In [8]:
random_state = 123
np.random.seed(random_state)
rand_matrix = np.random.normal(size=(4, 4))
rand_matrix

array([[-1.0856306 ,  0.99734545,  0.2829785 , -1.50629471],
       [-0.57860025,  1.65143654, -2.42667924, -0.42891263],
       [ 1.26593626, -0.8667404 , -0.67888615, -0.09470897],
       [ 1.49138963, -0.638902  , -0.44398196, -0.43435128]])

Проверим полученную матрицу rand_matrix на обратимость.

In [9]:
rand_matrix_inv = np.linalg.inv(rand_matrix)
rand_matrix_inv

array([[-0.43877556,  0.08607718, -1.66796984,  1.80033411],
       [-0.48457712,  0.30483394, -2.55861755,  1.93735555],
       [-0.10355901, -0.24183861, -1.21619062,  0.86313132],
       [-0.68794478,  0.09436506, -0.72043156,  0.14735665]])

Обратная матрица успешно получена, следовательно матрица rand_matrix действительно обратимая.

Умножим матрицу признаков на обратимую матрицу и сохраним, полученный результат.

In [10]:
features_new = features.dot(rand_matrix)

Проверим качество линейной регрессии при использовании исходной матрицы признаков. 

In [11]:
model = LinearRegression()
model.fit(features, target)
predictions = model.predict(features)
mse = mean_squared_error(target, predictions)
r2 = r2_score(target, predictions)
print("Средняя квадратичная ошибка модели равна:", mse)
print("Коэффициент детерминации модели равен:", r2)

Средняя квадратичная ошибка модели равна: 0.12527263826681606
Коэффициент детерминации модели равен: 0.4302010044852066


Проверим качество линейной регрессии при использовании измененной матрицы признаков. 

In [12]:
model.fit(features_new, target)
predictions = model.predict(features_new)
mse = mean_squared_error(target, predictions)
r2 = r2_score(target, predictions)
print("Средняя квадратичная ошибка модели равна:", mse)
print("Коэффициент детерминации модели равен:", r2)

Средняя квадратичная ошибка модели равна: 0.125272638266816
Коэффициент детерминации модели равен: 0.4302010044852068


Качество модели не изменилось. Таким образом, алгоритм преобразования данных работает успешно.

### Вывод 

Качество модели при использовании предложенного алгоритма преобразования данных с целью их защиты не изменилось. Таким образом, данный алгоритм работает успешно.

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования